library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,53 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - Every type is ABI-compatible with itself.
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`, `NonNull<T>` are all ABI-compatible with each
+///   other for all `T`.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
+/// - Any two 1-ZST types (types with size 0 and alignment 1) are ABI-compatible.
+/// - A `repr(transparent)` type `T` is ABI-compatible with its unique non-1-ZST field (if there is
+///   such a field).
+/// - `i32` is ABI-compatible with `NonZeroI32`, and similar for all other integer types with their
+///   matching `NonZero*` type.
+/// - If `T` is guaranteed to be subject to the [null pointer
+///   optimization](option/index.html#representation), then `T` and `Option<T>` are ABI-compatible.
+/// - If `T1` and `T2` are ABI-compatible, then two `repr(C)` types that only differ because one
+///   field type was changed from `T1` to `T2` are ABI-compatible.
+/// - ABI-compatibility is symmetric and transitive.
```

Comment:
i would maybe move this up to before this list of rules or to be the first bullet point. the rules logically get more specific as we descend but this seems the most general of all.

Reply:
I deliberately put it last; I think symmetry and transitivity together with the previous rule (which is the congruence rule) are the 3 closure properties that are less interesting than the rest of them -- the rest is basically "seeding" the relation with the actually interesting base cases; here we just complete the paperwork.

Reply:
Looking at the entire text again, I decided to separate these closure properties from the "seeds". Hopefully that's all more clear niw.